servicetalk-http-utils/src/main/java/io/servicetalk/http/utils/TimeoutHttpRequesterFilter.java

```diff
@@ -132,57 +137,38 @@ public TimeoutHttpRequesterFilter(final TimeoutFromRequest timeoutForRequest,
-                // No influence since we do not block.
-                return strategy;
-            }
-        };
-    }
-
     private Single<StreamingHttpResponse> request(final StreamingHttpRequester delegate,
                                                   final HttpExecutionStrategy strategy,
                                                   final StreamingHttpRequest request) {
         return Single.defer(() -> {
-            Single<StreamingHttpResponse> response = delegate.request(strategy, request);
-
             Duration timeout = timeoutForRequest.apply(request);
-
-            if (null != timeout) {
-                Single<StreamingHttpResponse> timeoutResponse = timeoutExecutor != null ?
-                        response.timeout(timeout, timeoutExecutor) : response.timeout(timeout);
-
-                response = fullRequestResponse ?
-                        Single.defer(() -> {
-                            Instant beganAt = Instant.now();
-                            return timeoutResponse.map(resp -> resp.transformMessageBody(body -> Publisher.defer(() -> {
-                                Duration remaining = timeout.minus(Duration.between(beganAt, Instant.now()));
-                                return (Duration.ZERO.compareTo(remaining) >= 0 ?
-                                        Publisher.failed(
-                                                new TimeoutException("timeout after " + timeout.toMillis() + "ms"))
-                                        : (null != timeoutExecutor ?
-                                        body.timeoutTerminal(remaining, timeoutExecutor)
-                                        : body.timeoutTerminal(remaining))
-                                ).subscribeShareContext();
-                            }))).subscribeShareContext();
-                        })
-                        : timeoutResponse;
+            Single<StreamingHttpResponse> response;
+            if (null != timeout && Duration.ZERO.compareTo(timeout) >= 0) {
+                response = Single.failed(new TimeoutException("negative timeout of " + timeout.toMillis() + "ms"));
+            } else {
+                response = delegate.request(strategy, request);
+
+                if (null != timeout) {
+                    Single<StreamingHttpResponse> timeoutResponse = timeoutExecutor == null ?
+                            response.timeout(timeout) : response.timeout(timeout, timeoutExecutor);
+
+                    if (fullRequestResponse) {
+                        Instant deadline = Instant.now().plus(timeout);
+                        response = timeoutResponse.map(resp -> resp.transformMessageBody(body ->
+                                Publisher.defer(() -> {
+                                    Duration remaining = Duration.between(Instant.now(), deadline);
+                                    return (Duration.ZERO.compareTo(remaining) < 0 ?
```

Comment:
Should it be `<= 0` here and in service filter?

Reply:
Indeed